codice/ddf
# DDF-5614 create CachePutPlugin interface and update SolrCache to call registered plugins

## What does this PR do?

Adds a CachePutPlugin and updates SolrCache to utilize the plugin. This plugin is called for every metacard being put into the solr cache. The plugin provides the opportunity to modify metacards before being put into the cache, or to indicate that the metacard should not be put into the cache.

This PR also provides an implementation of the CachePutPlugin that sets metacard tags attribute to "resource" if the metacard does not have a metacard tags.

#### Who is reviewing it? 


@pklinef 
@jlcsmith 
@jrnorth 

#### Select relevant component teams: 
@codice/core-apis 
@codice/data 
@codice/solr 

#### Ask 2 committers to review/merge the PR and tag them here.
@bdeining
@bdthomson

#### How should this be tested?
1. Setup GEOSERVER
2. Configure DDF to federate to GEOSERVER
3. Query against the GEOSERVER. Note that records are returned.
4. Shutdown GEOSERVER
5. Re-run the query. The query should return the metacards that were cached in Step 3.

#### Any background context you want to provide?

#### What are the relevant tickets?
Fixes: #__https://github.com/codice/ddf/issues/5614__

#### Screenshots


#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
